### ISSUE TITLE
BUG: MKL2020 Numpy.linalg breaks importing of DLLs for other packages in Windows

### ISSUE THREAD

**Issue by tristanic** [numpy/numpy#15431]
In Windows, using numpy.linalg.inv clears the application binary directory from the DLL search path, breaking future imports of any other library with DLLs there.

### Reproducing code example:

```python
import numpy as np
import win32api
d = win32api.GetDllDirectory()
print(win32api.GetDllDirectory())
# returns path to directory containing Python executable
arr = np.identity(3)
print(win32api.GetDllDirectory())
# Still fine
np.linalg.inv(arr)
print(win32api.GetDllDirectory())
# error: (0, 'GetDllDirectory', 'No error message is available')
```

### Error message:
error: (0, 'GetDllDirectory', 'No error message is available')

Attempting to subsequently import a compiled library requiring DLLs in the application base directory leads to e.g.

```python
from simtk import openmm

ImportError                               Traceback (most recent call last)
<ipython-input-4-9cb80959ee4f> in <module>
----> 1 from simtk import openmm

C:\Program Files\ChimeraX-Daily\bin\lib\site-packages\simtk\openmm\__init__.py in <module>
     17         'lib': version.openmm_library_path, 'path': _path}
     18 
---> 19 from simtk.openmm.openmm import *
     20 from simtk.openmm.vec3 import Vec3
     21 from simtk.openmm.mtsintegrator import MTSIntegrator

C:\Program Files\ChimeraX-Daily\bin\lib\site-packages\simtk\openmm\openmm.py in <module>
     26                 fp.close()
     27             return _mod
---> 28     _openmm = swig_import_helper()
     29     del swig_import_helper
     30 else:

C:\Program Files\ChimeraX-Daily\bin\lib\site-packages\simtk\openmm\openmm.py in swig_import_helper()
     22         if fp is not None:
     23             try:
---> 24                 _mod = imp.load_module('_openmm', fp, pathname, description)
     25             finally:
     26                 fp.close()

C:\Program Files\ChimeraX-Daily\bin\lib\imp.py in load_module(name, file, filename, details)
    240                 return load_dynamic(name, filename, opened_file)
    241         else:
--> 242             return load_dynamic(name, filename, file)
    243     elif type_ == PKG_DIRECTORY:
    244         return load_package(name, filename)

C:\Program Files\ChimeraX-Daily\bin\lib\imp.py in load_dynamic(name, path, file)
    341             name=name, loader=loader, origin=path)
    342         print('Name: {} Loader: {} origin: {}'.format(name, loader, path))
--> 343         return _load(spec)
    344 
    345 else:

ImportError: DLL load failed: The specified module could not be found.
```

... and monitoring with Process Monitor confirms that this is due to the application base dir not being searched for the required DLL.

### Numpy/Python version information:

1.18.1 3.7.4 (tags/v3.7.4:e09359112e, Jul  8 2019, 20:34:20) [MSC v.1916 64 bit (AMD64)]

**Comment by tristanic** [comment]
This doesn't occur in Numpy 1.17.4.

**Comment by charris** [comment]
How did you install NumPy? Are you using Anaconda or some other distribution?

**Comment by charris** [comment]
I believe there has been a change in the handling of DLLs due to recommendations from Microsoft.

**Comment by tristanic** [comment]
I’m using the Numpy bundled into UCSF ChimeraX. They updated from 1.17 a week or two ago. Copying their team in - they should be able to provide details.

**Comment by tristanic** [comment]
A comment in the ChimeraX numpy Makefile says for Windows

	# Downloaded from http://www.lfd.uci.edu/~gohlke/pythonlibs/ <http://www.lfd.uci.edu/~gohlke/pythonlibs/>

	https://www.rbvi.ucsf.edu/trac/ChimeraX/browser/prereqs/numpy <https://www.rbvi.ucsf.edu/trac/ChimeraX/browser/prereqs/numpy>

and the wheel file being used is named

	numpy-1.18.1+mkl-cp37-cp37m-win_amd64.whl

  Tom

**Comment by charris** [comment]
That's not one of our wheels, might be worth bringing @cgohlke in on the conversation. I suspect that the problem here is on the ChimeraX end but there really isn't enough information yet to draw conclusions.

**Comment by cgohlke** [comment]
This happens if anyone calls `SetDllDirectory(NULL)`, e.g.:
```
>>> import win32api
>>> win32api.SetDllDirectory('C:\\')
>>> print(win32api.GetDllDirectory())
C:\
>>> win32api.SetDllDirectory(None)
>>> print(win32api.GetDllDirectory())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pywintypes.error: (0, 'GetDllDirectory', 'No error message is available')
```

numpy-1.18.1+mkl is compiled against Intel MKL 2020. It looks like the `mkl_code.dll` calls `SetDllDirectory`, although I haven't looked in detail with a debugger yet. The MKL 2019 version included in older numpy+mkl wheels does not call `SetDllDirectory`.

**Comment by cgohlke** [comment]
Here's a standalone example that fails by loading `mkl_core.dll`:
```Python
import sys, ctypes, win32api
path = sys.prefix + '/Lib/site-packages/numpy/DLLs'
win32api.SetDllDirectory(path)
print(win32api.GetDllDirectory())
ctypes.CDLL(path + '/mkl_core.dll')
print(win32api.GetDllDirectory())
```

**Comment by matthew-brett** [comment]
Would it be reasonable to describe this as a bug in MKL 2020?

Quoting from:

```
The SetDllDirectory function affects all subsequent calls to the LoadLibrary and LoadLibraryEx functions.  It also effectively disables safe DLL search mode while the specified directory is in the search path.
```

It only allows one directory to be set at a time, so it's a very dangerous piece of global state.

**Comment by cgohlke** [comment]
> Would it be reasonable to describe this as a bug in MKL 2020?

Yes, I think so. Are any developers of Intel's Python team hanging out here? Otherwise I contact Intel support.

**Comment by charris** [comment]
@oleksandr-pavlyk Ping.

**Comment by cgohlke** [comment]
I raised the [issue on the Intel Support Forum](https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/844125).

**Comment by tristanic** [comment]
Thanks for the great detective work!

**Comment by oleksandr-pavlyk** [comment]
I tried to reproduce the problem on a Windows server (`conda create -n test python=3.7 pywin32`):

```
Python 3.7.6 (default, Jan  8 2020, 20:23:39) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import win32api
>>> win32api.GetDllDirectory()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pywintypes.error: (0, 'GetDllDirectory', 'No error message is available')
```

and using conda-forge stack (`conda create -n test2 -c conda-forge python=3.7 pywin32`):

```
Python 3.7.6 | packaged by conda-forge | (default, Jan  7 2020, 21:48:41) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import win32api
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: DLL load failed: The specified procedure could not be found.
```

Could you help me out by suggesting how to create an environment where `win32api.GetDllDirectory()` gets a non-error result?

**Comment by cgohlke** [comment]
> suggesting how to create an environment where win32api.GetDllDirectory() gets a non-error result?

For `GetDllDirectory()` to succeed you need to first call `SetDllDirectory()` with a valid path.

**Comment by oleksandr-pavlyk** [comment]
Thank you. Reproduced:

```python
import os, win32api
win32api.SetDllDirectory(os.environ['CONDA_PREFIX'] + '\\Lib')
win32api.GetDllDirectory()
# output 'C:\\devel\\miniconda3\\envs\\idp_test\\Lib'
import mkl
mkl.get_version_string() # forces loading of the library
# outputs 'Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191125 for Intel(R) 64 architecture applications'
win32api.GetDllDirectory()
# Traceback (most recent call last):
#  File "<stdin>", line 1, in <module>
# pywintypes.error: (0, 'GetDllDirectory', 'No error message is available')
```

**Comment by oleksandr-pavlyk** [comment]
The MKL team was able to reproduce the issue and targets the fix for MKL 2020 update 1.

**Comment by oleksandr-pavlyk** [comment]
Given that this is not an issue in NumPy, I think it can be closed.

**Comment by charris** [comment]
@oleksandr-pavlyk Thanks for the follow through.

**Comment by cgohlke** [comment]
This is now fixed in MKL 2020.0.1
